src/c2goto/library/semaphore_lib.c

```diff
@@ -0,0 +1,106 @@
+#include <semaphore.h>
+
+#define __ESBMC_sem_lock_field(a) ((a).__lock)
+#define __ESBMC_sem_count_field(a) ((a).__count)
+#define __ESBMC_sem_init_field(a) ((a).__init)
+
+/************************* Sem manipulation routines ************************/
+
+int sem_init(sem_t *__sem, int __pshared, unsigned int __value)
+{
+  __ESBMC_atomic_begin();
+  __ESBMC_assert(!__pshared, "ESBMC does not check multiprocesses for now");
+  __ESBMC_sem_lock_field(*__sem) = 0;
+  __ESBMC_sem_count_field(*__sem) = __value;
```

Comment:
This can overflow, in which case there would be undefined behaviour, which the `--overflow-check` should catch, but it's optional. In case we do not check for overflows the semaphore's value would end up negative, which should never happen, if I understand correctly. Could you add an assertion that `value <= INT_MAX` or similar?